Add assumptions on indexed symbols

bsamseth opened the pull request:
#### Brief description of what is fixed or changed

Constructor of ``IndexedBase`` now accepts assumptions the same way that ``Symbol`` accepts them. 

Fixes #13992.

#### Other comments

The following assertions all succeed after this change:

``` python
from sympy import Symbol, log, exp, ask, Q, IndexedBase, Indexed

i = Symbol("i", integer=True)
X = IndexedBase("X", real=True)

assert ask(Q.real(X))
assert ask(Q.real(X[i]))
assert log(exp(X)) == X
assert log(exp(X[i])) == X[i]
```

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* tensor
  * added support for assumptions on IndexedBase
<!-- END RELEASE NOTES -->

bsamseth committed:
Make assumptions about indexed symbols (#13992)

``Indexed`` and ``IndexedBase`` now accepts a new keyword argument
allowing the specification of assumptions on indexed symbols.

bsamseth committed:
Store copy of assumption.

Logic similar to that of ``Symbol``, simply copied.

bsamseth committed:
Extract assumption sanitizing to function

asmeurer commented on sympy/tensor/indexed.py:
```diff
@@ -137,7 +139,19 @@ class Indexed(Expr):
+        obj._assumptions = StdFactKB(assumptions)
+        obj._assumptions._generator = tmp_asm_copy  # Issue #8873
+
+    def __new__(cls, base, *args, assumptions=None, **kw_args):
```
This is not valid syntax in Python 2. You need to use `kw_args.setdefault('assumptions', None)`.

asmeurer commented:
Please add test cases to the tests. You should test using the old assumptions syntax (`X[i].is_real`), not `ask`.

asmeurer commented:
Why do you have it as `IndexedBase("X", assumptions={"real": True})` instead of `IndexedBase("X", real=True)`?

bsamseth commented:
@asmeurer Thanks for review this!

> Why do you have it as `IndexedBase("X", assumptions={"real": True})` instead of `IndexedBase("X", real=True)`?

I wanted to do this like you suggested, to follow the API of ``Symbol``. But I wasn't sure how to separate out assumption keywords from the other ``kw_args`` that are forwarded to ``Expr.__new__``. If, however, the ``kw_args`` are not used in practice, I could simply take this as ``**assumptions`` similarly to ``Symbol``. 

Another option might be to filter ``kw_args`` against this list: https://github.com/sympy/sympy/blob/d44e6038bec476750bcbaf0fcffff4a108097e3e/sympy/assumptions/ask.py#L1424-L1427

I'll start implementing this last option and adding test cases accordingly, and change direction depending on what you think.

bsamseth committed:
Make assumptions kw_args and not as part of dict

Previously:
    ``IndexedBase('A', assumptions={'real': True})``
This commit:
    ``IndexedBase('A', real=True)``

bsamseth committed:
Add assumptions0 property to Indexed(Base)

bsamseth committed:
Add test for Indexed(Base) assumptions

bsamseth committed:
Fix issue with assumptions forwarding from IndexedBase

Only forwarding the user-specified assumptions. Forwarding
after having passed them through ``StdFactKB`` leads to some
derived assumptions not being filtered out from the kw_args that
are sent to ``Expr.__new__``. This leads to an error with unexpected
keyword arguments.

bsamseth committed:
Indexed(Base)s with different assumptions are not equal

Equality now depends also on what assumptions are set, and
non-equivalent assumptions leads to `==` failing.

bsamseth committed:
Add assumption based equalities for Indexed(Base)

bsamseth committed:
Use Python 2 style super

bsamseth committed:
Ensure base sees same assumptions as Indexed ctor

bsamseth committed:
Ensure IndexedBase sugar works with assumptions

bsamseth commented on sympy/tensor/tests/test_indexed.py:
```diff
@@ -203,6 +203,26 @@ def test_IndexedBase_shape():
+def test_IndexedBase_assumptions():
+    i = Symbol('i', integer=True)
+    a = Symbol('a')
+    A = IndexedBase(a, positive=True)
```
Keyword-based assumptions supported.

oscargus commented:
@Abdullahjavednesar Even though I would have liked to honour your review request, I have very little knowledge of the assumptions system. Bordering to not even having tried it...

Abdullahjavednesar commented:
Thanks for your response @oscargus, I was looking for someone who could actively take this PR ahead.

smichr commented on sympy/tensor/indexed.py:
```diff
@@ -137,6 +140,24 @@ class Indexed(Expr):
+    def _filter_assumptions(kw_args):
+        """Split the given dict into two parts: assumptions and not assumptions.
+           Keys are taken as assumptions if they correspond to a handler on ``Q``."""
+        assumptions = {k: v for k, v in kw_args.items() if hasattr(Q, k)}
```
If this is the same as Symbol, then shouldn't we use the assumptions that are used on Symbols?

I think those are `from sympy.core.assumptions import _assume_defined`. The only two there that are not in dir(Q) are 'polar' and 'noninteger'.

bsamseth replied:
That makes a lot of sense to me. I didn't find the `_assume_defined` set when I was looking for a way to do the filtering at first. See the latest commit.

bsamseth committed:
Filter out assumptions based on _assume_defined

smichr commented on sympy/tensor/indexed.py:
```diff
@@ -137,6 +140,24 @@ class Indexed(Expr):
+        is_commutative = fuzzy_bool(assumptions.get('commutative', True))
+        assumptions['commutative'] = is_commutative
+        obj._assumptions = StdFactKB(assumptions)
+        obj._assumptions._generator = tmp_asm_copy  # Issue #8873
```
this doesn't include the commutativity -- is that right?

bsamseth replied:
Not quite sure what you mean - what do you mean it doesn't include the commutativity?  The logic here is based on `Symbol`, which defaults to be commutative if not specified otherwise. This does the same thing, I believe.

smichr replied:
`assumptions` is potentially changed at line 157 but you store, in `_generator`, the copy, not the updated assumptions.

smichr replied:
But I confirm that this is the same as in Symbol.

bsamseth replied:
Aha, I see what you mean, yes. Seems to me that the copy is intentionally made before the change, so as to detect what was specified by the user vs. what the end result was. At least that's how I understand #8873.

smichr commented:
Just one question, otherwise looks good.

smichr commented on sympy/tensor/tests/test_indexed.py:
```diff
@@ -203,6 +203,26 @@ def test_IndexedBase_shape():
+        assert log(exp(c)) == c
+
+    assert A != IndexedBase(a)
+    assert A == IndexedBase(a, positive=True, real=True)
```
is `real` redundant here?

bsamseth replied:
Yes, kind of, but on purpose. I left it in to verify that the assumptions were transformed into the same set, even if a user specifies redundant assumptions. Maybe unnecessary/unclear. Would you like me to change it? Perhaps as a separate test instead?

smichr replied:
it's ok

smichr commented on sympy/tensor/tests/test_indexed.py:
```diff
@@ -203,6 +203,26 @@ def test_IndexedBase_shape():
+def test_IndexedBase_assumptions():
+    i = Symbol('i', integer=True)
+    a = Symbol('a')
+    A = IndexedBase(a, positive=True)
```
I think we should not be encouraging use of Indexed. The docstring for it should be changed from 
```
It is recommended that ``Indexed`` objects be created via ``IndexedBase``
```
to
```
It is recommended that ``Indexed`` objects be created by indexing ``IndexedBase``:
IndexedBase('A')[i, j] instead of Indexed(IndexedBase('A'), i, j).
```
This class is created when IndexedBase has a `getitem` applied in the same way that `Function('f')` which is an Undef becomes AppliedUndef when it is called as `Function('f')(x)`. Functions are declared real at the Function level (`f = Function('f', real=True)` and then `f(x).is_real` -> True irrespective of what assumptions that `x` has in the same way that `f.is_real` -> True). 

In similar manner, I would expect that for `A = IndexedBase('A', real=True)` then we would have `A.is_real` and `A['i'].is_real`->True. A natural way to avoid having to split keywords is to require that IndexedBase inherits assumptions from the Symbol, if provided. We could do the same for Function, i.e. Function(Symbol('f', real=True)).is_real -> True. In such cases, however, I would raise an error if assumptions were passed when a Symbol was used: if the user doesn't want the Symbol's assumptions they could pass Symbol.name.

The main-point summary is that I don't think we should have `Indexed(IndexedBase('A'), i, j, real=True)`. The assumptions should be associated with IndexedBase, optionally inherited from the Symbol used to provide the name (but if that is done it should probably be done with Function, too).

smichr replied:
Instantiation of `IndexBase` could go as
```python
base = sympify(base)
assumptions = base.assumptions0
passed_assumptions, kwargs = split assumptions
sanitize passed_assumptions
for k, v in passed_assumptions.items():
    if v is None:  # can it be?
        continue
    vis = assumption.get(k, None)
    if vis is None:
        assumptions[k] = v
    elif v != vis:
        raise ValueError('clash between base and passed assumptions')
    else:
        pass  # v == vis
# assumptions are ready to store
```

bsamseth replied:
To be clear, you want to remove support for assumptions to be specified directly in the constructor of `Indexed` (with appropriate changes to the docstring), so that the only way to add assumptions is to set it on the `IndexedBase` object or inherit from the symbol used to instantiate `IndexedBase`? I'll be happy to do that, but just want to make sure about the desired changes. 

Should the part about inheriting from symbol be included in this PR, or should another one be made for adding this to both `Function` and `IndexedBase`? Could change `Function` in this PR also, but that seems less clear.

PS: Sorry about the delayed response.

smichr replied:
Yes, "only way to add assumptions is to set it on the IndexedBase".

And it makes sense to make inheritance from a Symbol a separate PR, handling Function at the same time.

bsamseth replied:
Done. I'll open a new PR for inheritance from Symbol if/when this is accepted.

bsamseth committed:
Remove direct support for assumptions on ``Indexed``

The only way to specify assumptions is to associate them to a ``IndexedBase``
object. Assumptions will be passed on to ``Indexed`` objects when created with
``__get_item_``. Meant to discourage direct use of ``Indexed``.

bsamseth committed:
Update IndexedBase docstring with assumptions example

bsamseth committed:
Fix trailing whitespace issue

smichr commented on sympy/tensor/indexed.py:
```diff
@@ -407,13 +451,23 @@ def __new__(cls, label, shape=None, **kw_args):
+        return {k: v for k, v in self._assumptions.items() if v is not None}
+
     def __getitem__(self, indices, **kw_args):
+        # Propagate assumptions onto new Indexed object.
```
this line should now be deleted?

bsamseth committed:
Remove outdated comment

smichr commented on sympy/tensor/indexed.py:
```diff
@@ -106,9 +106,12 @@
 
 from __future__ import print_function, division
 
+from sympy.assumptions import Q
```
unused Q import can be deleted

bsamseth committed:
Remove unused import

smichr commented on sympy/tensor/indexed.py:
```diff
@@ -156,7 +163,12 @@ def __new__(cls, base, *args, **kw_args):
                 return base[args]
 
-        return Expr.__new__(cls, base, *args, **kw_args)
+        obj = Expr.__new__(cls, base, *args, **kw_args)
```
Were you going to wait on this feature (garbbing assumptions from the base) and do Function and IndexedBase at the same time?

bsamseth replied:
That was the plan, yes. I left `Indexed` inheriting from `IndexedBase` in there, thinking that still made sense to include here, but I realize that inadvertently also inherits from `Symbol` into `Indexed`...  I still think it makes sense to have the currently suggested feature, but the question is whether or not it should be introduced in this PR or the second one also fixing `Function` and `IndexedBase`. Or at this point, I could just add everything and update the PR description accordingly? Sorry about the mess.

smichr replied:
> a distinct behavior

totally agree.

Please double check my edits of the OP (and edit at will). Otherwise this looks ready to go.

bsamseth replied:
The edits look good to me, no objections 👍

bsamseth committed:
Update failing doctest

The produced assumptions are different on the PR build on Travis compared to
this development branch, which results in a mismatch. The new edit shows the
existence of assumptions in a simpler manner which should resolve the issue.

bsamseth committed:
Propagate assumptions onto Indexed only if available

Tests from other parts of the code base invokes Indexed constructor with types
that do not have any assumptions support. These cases should be handled
gracefully.

Similarly, if no assumptions can be forwarded, we should take care to always set
the defaults. This way we don't have any ill-defined states where `Indexed`
objects fail if queried for their assumptions.

Note: Assumptions set on Symbols will not currently be forwarded to `Indexed`
objects. Only if `IndexedBase` grabs he assumptions first will they be passed along.
